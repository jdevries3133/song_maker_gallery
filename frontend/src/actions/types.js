// public
export const GET_GALLERY = "GET_GALLERY";

// user
export const GET_GALLERIES = "GET_GALLERIES";
export const POST_GALLERY = "POST_GALLERY";
export const DELETE_GALLERY = "DELETE_GALLERY";

// auth
export const LOGIN = "LOGIN";
export const LOGOUT = "LOGOUT";
export const REGISTER = "REGISTER";
export const CLEAR_ERROR = "CLEAR_ERROR";
export const UPDATE = "UPDATE";
export const DELETE = "DELETE";

// service notice
export const DISMISS_NOTIFICATION = "DISMISS_NOTIFICATION";
