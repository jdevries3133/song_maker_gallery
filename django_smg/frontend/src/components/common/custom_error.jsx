import React from "react";
import PropTypes from "prop-types";

import { H2 } from "./styles";
import { Blanket } from "./blanket";

/**
 * Light abstraction on Blanket to render an array of error messages.
 */
export const ErrorArray = (props) => {
  return (
    <Blanket onDismissed={props.onOk}>
      <H2 data-testid="errorArray header">{props.header}</H2>
      {props.message.map((msg, i) => (
        <p data-testid="errorArray message" key={i + msg}>
          {msg}
        </p>
      ))}
    </Blanket>
  );
};

ErrorArray.propTypes = {
  header: PropTypes.string,
  message: PropTypes.array.isRequired,
  onOk: PropTypes.func,
};

ErrorArray.defaultProps = {
  header: "Error",
};
