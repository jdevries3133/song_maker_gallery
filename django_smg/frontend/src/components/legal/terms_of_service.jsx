import React from "react";
import { Link } from "react-router-dom";

import styled, { Button, Div, Blanket, Description } from "Styles";

export const LegalContainer = styled(Div)`
  margin: 3%;
  margin-bottom: 5rem;
  & > * {
    text-align: justify;
  }
  & > * > * {
    margin-bottom: 1rem;
  }
  @media (max-width: 750px) {
    & > h1 {
      font-size: 36px;
    }
  }
`;

const TosContent = (props) => (
  <LegalContainer>
    <h1>Terms of Service</h1>
    <p>Updated 01/2021</p>
    <p>
      Also see our <Link to="/privacy/">privacy policy</Link>
    </p>
    <ol>
      <li>
        These terms constitute an agreement between the maintainer or
        representatives of this website, and any individual who has either
        registered an account or made a donation to support this website.
      </li>
      <li>
        This Agreement shall be governed by the laws of the State of New Jersey
      </li>
      <li>
        The user hereby acknowledges that they will not upload any personally
        identifying information of any individual under thirteen years of age as
        defined in the Children's Online Privacy Protection Rule ("COPPA"). The
        user furthermore agrees that they will not upload any personally
        identifying information of any minor whatsoever without the
        authorization to do so.
      </li>
      <li>
        By accepting this agreement, you acknowledge that you are at least 18
        years old.
      </li>
      <li>
        Any and all donations provided through PayPal to support this website
        are non-refundable. No expectation of goods, services, gifts, or
        benefits otherwise should be expected upon the submission of a donation
        through PayPal. Any premium features of this site must be purchased
        explicitly, and will not be rendered as the result of a PayPal donation.
      </li>
      <li>
        Users should be aware that this website is the hobby project of an
        individual who is neither a full-time web developer, nor a professional
        software engineer, nor a lawyer. The sole maintainer of this website is
        a music teacher with no amateur knowledge of many of the components that
        comprise this website. As such, all users (including users who have paid
        for premium features), should hold a reasonable expectation of site
        outage and service inavalibility.
      </li>
      <li>
        Refunds for any premium service which this site provides will be
        provided at the sole discretion of the maintainer of this website. In
        the case that a software malfunction, site outage, or other unplanned
        mistake cause the prolonged and significant interruption to a service
        that the user has paid for, the maintainer of this site
        <b>will make a reasonable effort to provide a refund.</b>
        HOWEVER, the user should note the fifth term of this agreement prior to
        purchasing any service that this website provides and consider the
        possibility that
        <b>
          even in the case of a prolonged service interruption, it is possible
          that this website can not or will not provide a refund.
        </b>
      </li>
      <li>
        The website is provided "as is without warranty of any kind, express or
        implied, including but not limited to the warranties of merchantability,
        fitness for a particular purpose and noninfringement. In no event shall
        the authors or copyright holders be liable for any claim, damages or
        other liability, whether in an action of contract, tort or otherwise,
        arising from, out of or in connection with the website or the use or
        other dealings in the website.
      </li>
      <li>
        User accounts may be terminated at any time for any reason. Likewise,
        individuals may be blocked from this website at any time for any reason
        and by any means.
      </li>
    </ol>
    <p>
      {" "}
      <a href="mailto:songmakergallery@gmail.com">Contact Us</a> with any
      questions or concerns about this policy.
    </p>
    {props.okButton ? props.okButton : null}
  </LegalContainer>
);

/* Free-standing page content */
export const TosPage = () => {
  return (
    <Description>
      <TosContent />
    </Description>
  );
};

/* modal with close button */
export const TosText = ({ onOk }) => {
  return (
    <Blanket>
      <TosContent
        okButton={
          <Button color="salmon" onClick={() => onOk()}>
            Close
          </Button>
        }
      />
    </Blanket>
  );
};
