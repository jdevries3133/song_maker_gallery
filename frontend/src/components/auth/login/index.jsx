import React, { useEffect, useState } from "react";
import { connect } from "react-redux";

import {
  login as loginAction,
  clearError,
  tryToken,
} from "Actions/auth.action";
import { getUserGalleries } from "Actions/user";
import { Redirect } from "react-router-dom";

import styled, { Div, H1 as DefaultH1 } from "Styles";
import { ErrorHandler } from "./error_handler";
import { LoginFooter } from "./login_footer";
import { LoginForm } from "./form";
import { windowLocation } from "../../../util/window";

const H1 = styled(DefaultH1)`
  margin-bottom: 0;
`;

const login = (props) => {
  const [username, updateUsername] = useState("");
  const [password, updatePassword] = useState("");
  const params = new URLSearchParams(windowLocation("search"));
  const token = localStorage.getItem("token");

  useEffect(() => {
    if (token) {
      props.tryToken(token);
    }
  }, []);

  if (props.isAuthenticated) {
    return <Redirect to={params.get("next") ?? "/teacher"} />;
  }

  const submit = (e) => {
    e.preventDefault();
    props.loginAction({ username: username, password: password });
  };

  return (
    <Div>
      <ErrorHandler err={props.authError} />
      <H1>login!</H1>
      <LoginForm
        username={(u) => updateUsername(u)}
        password={(p) => updatePassword(p)}
        submit={submit}
      />
      <LoginFooter />
    </Div>
  );
};

const mapStateToProps = (state) => {
  return {
    authError: state.auth.authError,
    isAuthenticated: state.auth.isAuthenticated,
    galleries: state.user.galleries,
  };
};

export default connect(mapStateToProps, {
  loginAction,
  clearError,
  getUserGalleries,
  tryToken,
})(login);
