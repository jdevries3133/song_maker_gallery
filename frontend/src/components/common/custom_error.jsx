import React from "react";
import PropTypes from "prop-types";

import { Blanket } from "Common/blanket";
import { H2 } from "./styles";

/**
 * Render an array of error messages
 */
export const ErrorArray = (props) => {
  return (
    <Blanket onDismissed={props.onOk}>
      <H2 data-testid="errorArray header">{props.header}</H2>
      {props.message.map((par, i) => (
        <p data-testid="errorArray message" key={i}>
          {par}
        </p>
      ))}
    </Blanket>
  );
};

ErrorArray.propTypes = {
  header: PropTypes.string,
  message: PropTypes.array.isRequired,
  onOk: PropTypes.func.isRequired,
};

ErrorArray.defaultProps = {
  header: "Error",
};
