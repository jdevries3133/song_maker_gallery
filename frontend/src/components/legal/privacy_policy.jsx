import React from "react";
import { Link } from "react-router-dom";

import { Description } from "Styles";
import { Blanket } from "Common/blanket";
import { LegalContainer } from "./terms_of_service";

export const PrivacyContent = (props) => (
  <LegalContainer>
    <h1 id="privacy-policy">Privacy Policy</h1>
    <p>
      This Agreement shall be governed by the laws of the State of New Jersey
    </p>
    <p>
      Also see our <Link to="/tos/">Terms of Service</Link>
    </p>
    <ol>
      <li>
        Do not upload any personallly identifying information (as defined in the
        Children's Online Privacy Protection Rule ("COPPA")) of any individual
        under thirteen years of age.
      </li>
      <li>
        Do not upload the personally identifying information of any individual
        under eighteen years of age without authorization to do so from that
        individual's parent or guardian.
      </li>
    </ol>
    {props.okButton ? props.okButton : null}
  </LegalContainer>
);

export const PrivacyPage = () => (
  <Description>
    <PrivacyContent />
  </Description>
);

export const PrivacyText = () => (
  <Blanket>
    <div style={{ height: "2rem" }} />
    <PrivacyContent />
  </Blanket>
);
