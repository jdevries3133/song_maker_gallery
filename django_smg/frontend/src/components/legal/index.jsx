/**
 * terms of service
 */

import React, { useState, Fragment } from "react";
import styled from "styled-components";

import { TosText } from "./terms_of_service";
import { PrivacyText } from "./privacy_policy";

export { TosPage } from "./terms_of_service";
export { PrivacyPage } from "./privacy_policy";

/* Force <a> styling on anchor tags without href */
const A = styled.a`
  color: blue;
  font-weight: bold;
  text-decoration: underline;

  &:hover {
    cursor: pointer;
  }
`;

export const Tos = () => {
  const [showModal, setShowModal] = useState(false);
  const toggleModal = () => setShowModal(!showModal);
  return (
    <Fragment>
      {showModal ? <TosText onOk={toggleModal} /> : null}
      <A onClick={toggleModal}>Terms of Service</A>
    </Fragment>
  );
};

export const Privacy = () => {
  const [showModal, setShowModal] = useState(false);
  const toggleModal = () => setShowModal(!showModal);
  return (
    <Fragment>
      {showModal ? <PrivacyText onOk={toggleModal} /> : null}
      <A onClick={toggleModal}>Privacy Policy</A>
    </Fragment>
  );
};
