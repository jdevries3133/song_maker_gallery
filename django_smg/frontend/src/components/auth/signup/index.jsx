import React, { useState } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";

import { register, clearError } from "Actions/auth.action";
import styled, {
  Div,
  Input,
  P,
  Button,
  Label,
  Checkbox,
  Description,
} from "Styles";
import { ErrorArray } from "Common/custom_error";
import { Tos, Privacy } from "../../legal";

import { ValidationMessages } from "./validation_messages";

export const PASSWD_MIN_LENGTH = 9;

const Wrapper = styled.div`
  width: 100%;
  text-align: center;
`;

const Container = styled(Description)`
  max-width: 800px;
`;

// subtle adjust to align text with checkbox
const SpanCbAligned = styled.span`
  position: relative;
  bottom: 8px;
`;

const Header = styled.h1`
  margin: 0rem;
  font-size: 60px;
`;

const FlexForm = styled.form`
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  gap: 3rem;
`;

const SubmitButton = styled(Button)`
  margin-bottom: 1.5rem;
  font-size: 24px;
  height: auto;
  line-height: 40px;
  padding: 10px 5px 10px 5px;
  background-color: lightseagreen;
`;

const signup = (props) => {
  const [emailInput, updateEmail] = useState("");
  const [usernameInput, updateUsername] = useState("");
  const [passwordInput, updatePassword] = useState("");
  const [passwordConfirm, updateConfirm] = useState("");
  const [blanket, setBlanket] = useState(null);
  const clearBlanket = () => setBlanket(null);
  const [TOS, setTOS] = useState(false);

  /**
   * Validate form and inject a modal warning into "blanket" if it is not
   * valid. Checks for the following:
   *
   * - All fields are filled
   * - Password and confirm password are the same
   * - Terms of service box has been checked
   *
   * @returns {bool} whether form is valid
   */
  const validate = () => {
    if (
      emailInput === "" ||
      usernameInput === "" ||
      passwordInput === "" ||
      passwordConfirm === ""
    ) {
      setBlanket(
        <ErrorArray
          header="Blank Fields"
          message={["Required fields are blank", "All fields are required."]}
          onOk={clearBlanket}
        />
      );
      return false;
    }
    if (passwordInput !== passwordConfirm) {
      setBlanket(
        <ErrorArray
          header="Passwords do not match"
          message={[""]}
          onOk={clearBlanket}
        />
      );
      return false;
    }
    if (!TOS) {
      setBlanket(
        <ErrorArray
          header="Terms of Service"
          message={["You must accept the terms of service to make an account"]}
          onOk={clearBlanket}
        />
      );
      return false;
    }
    return true;
  };

  const submit = (e) => {
    e.preventDefault();
    if (validate()) {
      props.register({
        email: emailInput,
        username: usernameInput,
        password: passwordInput,
      });
    }
  };

  if (props.isAuthenticated) {
    return <Redirect to="/teacher" />;
  } else if (props.authError && !blanket) {
    const errors = Object.keys(props.authError).map(
      (k) => `${k}: ${props.authError[k]}`
    );
    setBlanket(
      <ErrorArray
        onOk={() => {
          props.clearError();
          clearBlanket();
        }}
        header="Validation Error"
        message={errors}
      />
    );
  }

  return (
    <Wrapper>
      {blanket}
      <Header>sign up!</Header>
      <Container>
        <FlexForm data-testid="signupForm" onSubmit={submit}>
          <Div>
            <Label htmlFor="email">Email</Label>
            <Input
              id="email"
              autoComplete="email"
              data-testid="emailInput"
              onChange={(event) => updateEmail(event.target.value)}
            />
            <Label htmlFor="username">Username</Label>
            <Input
              id="username"
              autoComplete="username"
              data-testid="usernameInput"
              style={
                usernameInput.includes(" ") ? { borderColor: "red" } : null
              }
              value={usernameInput}
              onChange={(event) => updateUsername(event.target.value)}
            />
          </Div>
          <Div>
            <Label htmlFor="password">Password</Label>
            <Input
              id="password"
              autoComplete="new-password"
              data-testid="passwordInput"
              type="password"
              value={passwordInput}
              onChange={(event) => updatePassword(event.target.value)}
              style={
                passwordInput &&
                passwordConfirm &&
                passwordInput !== passwordConfirm
                  ? { borderColor: "red" }
                  : null
              }
            />
            <Label htmlFor="confirm password">Confirm Password</Label>
            <Input
              id="confirm password"
              autoComplete="new-password"
              data-testid="passwordConfirmInput"
              value={passwordConfirm}
              onChange={(e) => updateConfirm(e.target.value)}
              type="password"
              style={
                passwordInput &&
                passwordConfirm &&
                passwordInput !== passwordConfirm
                  ? { borderColor: "red" }
                  : null
              }
            />
          </Div>
          <Div>
            <ValidationMessages
              usernameInput={usernameInput}
              passwordInput={passwordInput}
              passwordConfirm={passwordConfirm}
            />
            <P>
              <SpanCbAligned>
                I agree to the <Tos /> and <Privacy />
                {"    "}
              </SpanCbAligned>
              <Checkbox
                id="tos"
                onClick={() => setTOS(!TOS)}
                data-testid="tosCheckbox"
              />
            </P>
            <Div>
              <SubmitButton as="input" type="submit" value="Sign Up" />
            </Div>
            <Div>
              <Link to="/login">
                <Button>Already have an account? Login here!</Button>
              </Link>
            </Div>
          </Div>
        </FlexForm>
      </Container>
    </Wrapper>
  );
};

const mapStateToProps = (state) => {
  return {
    authError: state.auth.authError,
    isAuthenticated: state.auth.isAuthenticated,
  };
};

export default connect(mapStateToProps, { register, clearError })(signup);
