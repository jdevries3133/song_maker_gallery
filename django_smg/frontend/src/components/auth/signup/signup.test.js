import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import {
  render,
  screen,
  fireEvent,
  act,
  waitFor,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import { Context } from "Test/app_context";

import Signup from "./index";
import { REGISTER } from "Actions/types";
import { register } from "Actions/auth.action";
import { Tos, Privacy } from "../../legal";
import { ErrorArray } from "Common/custom_error";

jest.mock("Common/custom_error");
jest.mock("../../legal");
jest.mock("Actions/auth.action");

Tos.mockImplementation(() => <span>TOS MOCK</span>);
Privacy.mockImplementation(() => <span>PRIVACY POLICY MOCK</span>);
ErrorArray.mockImplementation(() => <span>MOCK ERRORARRAY</span>);

beforeEach(() => {
  render(
    <Context>
      <Signup />
    </Context>
  );
});

afterEach(() => {
  register.mockClear();
  ErrorArray.mockClear();
});

const submitForm = async (
  data = {
    email: "test@testuser.com",
    username: "testuser",
    password: "123456789",
  }
) => {
  act(() => {
    fireEvent.change(screen.getByTestId("emailInput"), {
      target: { value: data.email },
    });
    fireEvent.change(screen.getByTestId("usernameInput"), {
      target: { value: data.username },
    });
    fireEvent.change(screen.getByTestId("passwordInput"), {
      target: { value: data.password },
    });
    fireEvent.change(screen.getByTestId("passwordConfirmInput"), {
      target: { value: data.password },
    });
    fireEvent.click(screen.getByTestId("tosCheckbox"));
  });
  act(() => {
    fireEvent.submit(screen.getByTestId("signupForm"));
  });
  return data;
};

describe("<Signup />", () => {
  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <Context>
          <Signup />
        </Context>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("tells user when passwd is too short", () => {
    expect(
      screen.getByText("Your password must be at least eight characters long")
    ).toBeVisible();
  });

  it("tells user when passswd is long enough with green text", async (done) => {
    // action: input password
    act(() => {
      fireEvent.change(screen.getByTestId("passwordInput"), {
        target: { value: "123456789" },
      });
    });

    // assertions
    await waitFor(() => {
      expect(
        screen.queryByText("Your password is at least eight characters long")
      ).toBeVisible();
      expect(
        screen.queryByText("Your password is at least eight characters long")
      ).toHaveStyle("color: green;");
    });
    done();
  });
  it("warns user if they put a space in the username", async (done) => {
    act(() => {
      fireEvent.change(screen.getByTestId("usernameInput"), {
        target: { value: " " },
      });
    });
    expect(
      screen.queryByText("Username may not contain spaces.")
    ).toBeVisible();
    act(() => {
      fireEvent.change(screen.getByTestId("usernameInput"), {
        target: { value: "nospaces" },
      });
    });
    expect(screen.queryByText("Username may not contain spaces.")).toBeNull();
    done();
  });
  it("restores original passwd length message after deleting password", async (done) => {
    // actions
    // input password
    act(() => {
      fireEvent.change(screen.getByTestId("passwordInput"), {
        target: { value: "123456789A" },
      });
    });
    // partially remove password
    act(() => {
      fireEvent.change(screen.getByTestId("passwordInput"), {
        target: { value: "123456" },
      });
    });

    // assertions
    await waitFor(() => {
      expect(
        screen.getByText("Your password must be at least eight characters long")
      ).toBeVisible();
    });
    done();
  });
  it("disallows empty fields", async (done) => {
    act(() => {
      fireEvent.submit(screen.getByTestId("signupForm"));
    });
    await waitFor(() => {
      expect(ErrorArray).toHaveBeenCalledWith(
        {
          header: "Blank Fields",
          message: ["Required fields are blank", "All fields are required."],
          onOk: expect.anything(),
        },
        expect.anything() // other args from react framework, etc.
      );
    });
    done();
  });
  it("calls user register action on form submission", async (done) => {
    const data = await submitForm();
    await waitFor(() => {
      expect(register).toHaveBeenCalledTimes(1);
      expect(register).toHaveBeenCalledWith(data);
    });
    done();
  });
  it("renders 400 error messages from server", async (done) => {
    const fakeApiResponse = {
      username: ["A user with that username already exists"],
      email: ["Invalid email address"],
    };
    register.mockImplementation(() => (dispatch) => {
      dispatch({
        type: REGISTER,
        payload: {
          isAuthenticated: false,
          authError: fakeApiResponse,
          token: null,
          user: null,
        },
      });
    });
    await submitForm();
    expect(ErrorArray).toHaveBeenCalledWith(
      {
        header: "Validation Error",
        message: expect.arrayContaining([
          `username: ${fakeApiResponse["username"]}`,
          `email: ${fakeApiResponse["email"]}`,
        ]),
        onOk: expect.anything(),
      },
      expect.anything()
    );
    done();
  });
});
