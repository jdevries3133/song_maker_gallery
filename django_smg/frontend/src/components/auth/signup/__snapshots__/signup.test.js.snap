// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> matches snapshot 1`] = `
.c5 {
  text-align: center;
  padding: inherit;
}

.c9 {
  text-align: center;
  padding: 0 20px 0 20px;
  color: green;
}

.c10 {
  text-align: center;
  padding: 0 20px 0 20px;
}

.c12 {
  display: inline-block;
  margin: auto;
  font-size: 1rem;
  font-weight: bold;
  padding: 20px;
  border-radius: 20px;
  cursor: pointer;
  background-color: #f7943e;
}

.c12:hover {
  background-color: white;
}

.c6 {
  display: block;
  font-size: 1.17em;
  margin-block-start: 1em;
  margin-block-end: 1em;
  margin-inline-start: 0px;
  margin-inline-end: 0px;
  font-weight: bold;
}

.c7 {
  max-width: 100%;
  font-size: 20px;
  padding: 10px;
  border-radius: 5px;
}

.c7[type="checkbox"] {
  width: 30px;
  height: 30px;
}

.c2 {
  display: inline-block;
  text-align: center;
  margin: 3vw;
  padding: 3vw;
  font-size: 18px;
  background: white;
  border-radius: 20px;
  box-shadow: 0px 3px 8px rgb(100,100,100);
}

.c8 {
  list-style-type: none;
  background-color: #adff1626;
  border-radius: 20px;
  padding: 1em;
}

.c0 {
  width: 100%;
  text-align: center;
}

.c3 {
  max-width: 800px;
}

.c11 {
  position: relative;
  bottom: 8px;
}

.c1 {
  margin: 0rem;
  font-size: 60px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 3rem;
}

.c13 {
  margin-bottom: 1.5rem;
  font-size: 24px;
  height: auto;
  line-height: 40px;
  padding: 10px 5px 10px 5px;
  background-color: lightseagreen;
}

@media (max-width:400px) {
  .c12 {
    padding: 10px;
  }
}

@media (max-width:350px) {
  .c12 {
    padding: 5px;
  }
}

@media (max-width:600px) {
  .c2 {
    font-size: 16px;
    line-height: 1.5rem;
  }
}

<div
  className="c0"
>
  <h1
    className="c1"
  >
    sign up!
  </h1>
  <div
    className="c2 c3"
  >
    <form
      className="c4"
      data-testid="signupForm"
      onSubmit={[Function]}
    >
      <div
        className="c5"
      >
        <label
          className="c6"
          htmlFor="email"
        >
          Email
        </label>
        <input
          autoComplete="email"
          className="c7"
          data-testid="emailInput"
          id="email"
          onChange={[Function]}
        />
        <label
          className="c6"
          htmlFor="username"
        >
          Username
        </label>
        <input
          autoComplete="username"
          className="c7"
          data-testid="usernameInput"
          id="username"
          onChange={[Function]}
          style={null}
          value=""
        />
      </div>
      <div
        className="c5"
      >
        <label
          className="c6"
          htmlFor="password"
        >
          Password
        </label>
        <input
          autoComplete="new-password"
          className="c7"
          data-testid="passwordInput"
          id="password"
          onChange={[Function]}
          style={null}
          type="password"
          value=""
        />
        <label
          className="c6"
          htmlFor="confirm password"
        >
          Confirm Password
        </label>
        <input
          autoComplete="new-password"
          className="c7"
          data-testid="passwordConfirmInput"
          id="confirm password"
          onChange={[Function]}
          style={null}
          type="password"
          value=""
        />
      </div>
      <div
        className="c5"
      >
        <ul
          className="c8"
        >
          <li
            className="c9"
          >
            Username does not contain any spaces
          </li>
          <li
            className="c10"
          >
            Your password must be at least eight characters long
          </li>
        </ul>
        <p
          className="c10"
        >
          <span
            className="c11"
          >
            I agree to the 
            <span>
              TOS MOCK
            </span>
             and 
            <span>
              PRIVACY POLICY MOCK
            </span>
                
          </span>
          <input
            className="c7"
            data-testid="tosCheckbox"
            id="tos"
            onClick={[Function]}
            type="checkbox"
          />
        </p>
        <div
          className="c5"
        >
          <input
            className="c12 c13"
            type="submit"
            value="Sign Up"
          />
        </div>
        <div
          className="c5"
        >
          <a
            href="/login"
            onClick={[Function]}
          >
            <button
              className="c12"
            >
              Already have an account? Login here!
            </button>
          </a>
        </div>
      </div>
    </form>
  </div>
</div>
`;
